README.md
# flutter_map_sdk_v1

CHANGELOG.md
## [0.0.5] - 9/23/2020.

* initial release.


lib/mapir.dart
import 'package:flutter/cupertino.dart';

class Mapir extends StatefulWidget {
  static String apiKey = "";
  static Mapir instance;
  static String userAgent = "";

  static Mapir init(context, apiKey) {
    if (instance == null)
      return Mapir(context, apiKey);
    else
      return instance;
  }

  Mapir(context, _apiKey) {
    apiKey = _apiKey;
    userAgent = userAgentString(context);
  }

  String userAgentString(context) {
    // if (Platform.isAndroid) {
    //   var androidInfo = DeviceInfoPlugin().androidInfo;
    //   return toHumanReadableAscii("Android/"
    //       "${androidInfo.version.sdkInt}"
    //       "(${androidInfo.version.release})"
    //       "(${androidInfo.supportedAbis})-FlutterSdk/"
    //       "22222222-"
    //       "${context.getPackageName()}");
    // } else if (Platform.isIOS) {
    //   var iosInfo = DeviceInfoPlugin().iosInfo;
    //
    //   return toHumanReadableAscii("Ios/"
    //       "${iosInfo.utsname.version}"
    //       "(${iosInfo.utsname.release})"
    //       "(${iosInfo.utsname.machine})-FlutterSdk/"
    //       "222222222-"
    //       "${context.getPackageName()}");
    // } else
    return "Neither Android Nor Ios";
  }

  static String toHumanReadableAscii(String s) {
    return s;
  }

  @override
  _MapirState createState() => _MapirState();
}

class _MapirState extends State<Mapir> {
  @override
  Widget build(BuildContext context) {
    return null;
  }
}


lib/MapView.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';
import 'package:latlong/latlong.dart';
import 'package:map_controller/map_controller.dart';

import 'mapir.dart';

class MapView extends StatefulWidget {
  StatefulMapController statefulMapController;
  MapController mapController;

  MapView({Key key, this.mapController, this.statefulMapController})
      : super(key: key);

  // StatefulMapController _statefulMapController;
  //
  // void setController(StatefulMapController controller) {
  //   _statefulMapController = controller;
  // }

  @override
  _MapViewState createState() => _MapViewState();
}

class _MapViewState extends State<MapView> {
  // MapController mapController;

  // StatefulMapController statefulMapController;
  StreamSubscription<StatefulMapControllerStateChange> sub;

  MapOptions mapOptions = MapOptions(
    center: LatLng(35.713949, 51.399086),
    zoom: 15,
    minZoom: 1.0,
    maxZoom: 18.0,
    // plugins: [PopupMarkerPlugin()],
    // onTap: (_) => popupLayerController.hidePopup()
  );

  // StatefulMapController getController() {
  //   return statefulMapController;
  // }

  @override
  void initState() {
    // wait for the controller to be ready before using it
    widget.statefulMapController.onReady
        .then((_) => print("The map controller is ready"));

    /// [Important] listen to the changefeed to rebuild the map on changes:
    /// this will rebuild the map when for example addMarker or any method
    /// that mutates the map assets is called
    sub = widget.statefulMapController.changeFeed
        .listen((change) => setState(() {}));

    super.initState();
  }

  @override
  void dispose() {
    sub.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
          child: Stack(children: <Widget>[
        FlutterMap(
          mapController: widget.mapController,
          options: mapOptions,
          layers: [
            TileLayerOptions(
              wmsOptions: WMSTileLayerOptions(
                  baseUrl: "https://map.ir/shiveh?"
                      "&EXCEPTIONS=application/vnd.ogc.se_inimage"
                      "&width=256"
                      "&height=256"
                      "&srs=EPSG:3857"
                      "&bbox=%f,%f,%f,%f&x-api-key=${Mapir.apiKey}",
                  layers: ['Shiveh:Shiveh'],
                  format: 'image/png',
                  version: '1.1.0',
                  otherParameters: {}),
            ),
            MarkerLayerOptions(markers: widget.statefulMapController.markers),
            PolylineLayerOptions(polylines: widget.statefulMapController.lines),
            PolygonLayerOptions(polygons: widget.statefulMapController.polygons)
          ],
        ),
        // ...
      ])),
    );
  }
}


lib/MapirMap.dart
import 'dart:io';

import 'package:device_info/device_info.dart';

class MapirMap {
  static String apiKey = "";
  static MapirMap instance;
  String userAgent = "";

  static MapirMap init(packageName, apiKey) {
    if (instance == null)
      return MapirMap(packageName, apiKey);
    else
      return instance;
  }

  MapirMap(packageName, _apiKey) {
    apiKey = _apiKey;
    userAgent = userAgentString(packageName);
  }

  String userAgentString(packageName) {
    if (Platform.isAndroid) {
      var androidInfo = DeviceInfoPlugin().androidInfo;
      return toHumanReadableAscii("Android/"
          //     "${androidInfo.version.sdkInt}"
          //     "(${androidInfo.version.release})"
          //     "(${androidInfo.supportedAbis})"
          "-FlutterSdk/"
          "0.0.1-"
          "$packageName");
    } else if (Platform.isIOS) {
      var iosInfo = DeviceInfoPlugin().iosInfo;
      return toHumanReadableAscii("Ios/"
          //       "${iosInfo.utsname.version}"
          //       "(${iosInfo.utsname.release})"
          //       "(${iosInfo.utsname.machine})-"
          "FlutterSdk/"
          "0.0.1-"
          "$packageName");
    } else
      return "Neither Android Nor Ios";
  }

  static String toHumanReadableAscii(String s) {
    return s;
  }
}


sample/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_map_sdk_v1/flutter_map_sdk_v1.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    MapirMap.init("نام پکیج برنامه شما", "کلید دسترسی");

    var mapController = MapController();
    var statefulMapController =
        StatefulMapController(mapController: mapController);

    var marker1 = Marker(
      width: 160.0,
      height: 160.0,
      point: LatLng(35.713949, 51.399086),
      builder: (ctx) => Container(child: Icon(Icons.location_on)),
      anchorPos: AnchorPos.align(AnchorAlign.center),
    );

    var points = List<LatLng>();
    points.add(LatLng(35.713949, 51.399086));
    points.add(LatLng(35.714345, 51.400304));
    points.add(LatLng(35.712942, 51.400486));

    statefulMapController.addLine(
        name: "sampleLine1",
        points: points,
        width: 5.0,
        color: Colors.blue,
        isDotted: true);

    statefulMapController.addPolygon(
        name: "samplePolygon1",
        points: points,
        color: Colors.blue,
        borderWidth: 5.0,
        borderColor: Colors.red);

    statefulMapController.addMarker(marker: marker1, name: "sampleMarker1");

    return MaterialApp(
        debugShowCheckedModeBanner: false,
        home: MapirMapView(
            mapController: mapController,
            statefulMapController: statefulMapController,
            onMapReady: () {},
            onMapClick: (location) {},
            onMapLongClick: (location) {},
            onLocationUpdate: (location) {},
            showLocationMarker: false));
  }
}


lib/marker_popup.dart
import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';

class MarkerPopup extends StatefulWidget {
  final Marker marker;

  MarkerPopup(this.marker, {Key key}) : super(key: key);

  @override
  State<StatefulWidget> createState() => _MarkerPopupState(this.marker);
}

class _MarkerPopupState extends State<MarkerPopup> {
  final Marker _marker;

  final List<IconData> _icons = [
    Icons.star_border,
    Icons.star_half,
    Icons.star
  ];
  int _currentIcon = 0;

  _MarkerPopupState(this._marker);

  @override
  Widget build(BuildContext context) {
    return Card(
      child: InkWell(
        child: Row(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Padding(
              padding: EdgeInsets.only(left: 20, right: 10),
              child: Icon(_icons[_currentIcon]),
            ),
            _cardDescription(context),
          ],
        ),
        onTap: () => setState(() {
          _currentIcon = (_currentIcon + 1) % _icons.length;
        }),
      ),
    );
  }

  Widget _cardDescription(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(10),
      child: Container(
        constraints: BoxConstraints(minWidth: 100, maxWidth: 200),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          mainAxisAlignment: MainAxisAlignment.start,
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Text(
              "Popup for a marker!",
              overflow: TextOverflow.fade,
              softWrap: false,
              style: const TextStyle(
                fontWeight: FontWeight.w500,
                fontSize: 14.0,
              ),
            ),
            const Padding(padding: EdgeInsets.symmetric(vertical: 4.0)),
            Text(
              "Position: ${_marker.point.latitude}, ${_marker.point.longitude}",
              style: const TextStyle(fontSize: 12.0),
            ),
            Text(
              "Marker size: ${_marker.width}, ${_marker.height}",
              style: const TextStyle(fontSize: 12.0),
            ),
          ],
        ),
      ),
    );
  }
}


lib/MapirMapView.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';
import 'package:latlong/latlong.dart';
import 'package:location/location.dart';
import 'package:map_controller/map_controller.dart';

import 'MapirMap.dart';

class MapirMapView extends StatefulWidget {
  StatefulMapController statefulMapController;
  MapController mapController;
  VoidCallback onMapReady;
  Function(LatLng) onMapClick;
  Function(LatLng) onMapLongClick;
  Function(LatLng) onLocationUpdate;
  LatLng lastKnownLocation;
  bool showLocationMarker;

  MapirMapView(
      {Key key,
      this.mapController,
      this.statefulMapController,
      this.onMapReady,
      this.onMapClick,
      this.onMapLongClick,
      this.onLocationUpdate,
      this.showLocationMarker = false})
      : super(key: key);

  @override
  _MapirMapViewState createState() => _MapirMapViewState();
}

class _MapirMapViewState extends State<MapirMapView> {
  StreamSubscription<StatefulMapControllerStateChange> sub;

  MapOptions mapOptions;

  @override
  void initState() {
    // wait for the controller to be ready before using it
    widget.statefulMapController.onReady.then((_) => widget.onMapReady());

    /// [Important] listen to the changefeed to rebuild the map on changes:
    /// this will rebuild the map when for example addMarker or any method
    /// that mutates the map assets is called
    sub = widget.statefulMapController.changeFeed
        .listen((change) => setState(() {}));

    mapOptions = MapOptions(
        center: LatLng(35.713949, 51.399086),
        zoom: 15,
        minZoom: 1.0,
        maxZoom: 18.0,
        onTap: (point) => widget.onMapClick(point),
        onLongPress: (point) => widget.onMapLongClick(point));

    if (widget.onLocationUpdate != null)
      getCurrentLocation(
          statefulMapController: widget.statefulMapController,
          showMarker: widget.showLocationMarker);
    else if (widget.showLocationMarker)
      getCurrentLocation(
          statefulMapController: widget.statefulMapController,
          showMarker: widget.showLocationMarker);

    super.initState();
  }

  @override
  void dispose() {
    sub.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
          child: Stack(children: <Widget>[
        FlutterMap(
          mapController: widget.mapController,
          options: mapOptions,
          layers: [
            TileLayerOptions(
              wmsOptions: WMSTileLayerOptions(
                  baseUrl: "https://map.ir/shiveh?"
                      "&EXCEPTIONS=application/vnd.ogc.se_inimage"
                      "&width=256"
                      "&height=256"
                      "&srs=EPSG:3857"
                      "&bbox=%f,%f,%f,%f&x-api-key=${MapirMap.apiKey}",
                  layers: ['Shiveh:Shiveh'],
                  format: 'image/png',
                  version: '1.1.0',
                  otherParameters: {}),
            ),
            MarkerLayerOptions(markers: widget.statefulMapController.markers),
            PolylineLayerOptions(polylines: widget.statefulMapController.lines),
            PolygonLayerOptions(polygons: widget.statefulMapController.polygons)
          ],
        ),
        Align(
            alignment: Alignment.bottomRight,
            child: Padding(
                child: Wrap(
                  children: <Widget>[
                    Text('\u00a9Map \u00a9OpenStreetMap',
                        style: TextStyle(height: 5, fontSize: 12)),
                  ],
                ),
                padding: EdgeInsets.fromLTRB(0, 0, 12, 0))),
        Align(
            alignment: Alignment.bottomLeft,
            child: Padding(
                child: Image.asset(
                    'packages/flutter_map_sdk_v1/assets/images/logo.png',
                    width: 100,
                    height: 48),
                padding: EdgeInsets.fromLTRB(8, 0, 0, 8)))
      ])),
    );
  }

  void getCurrentLocation(
      {StatefulMapController statefulMapController,
      bool showMarker = false}) async {
    var location = Location();

    bool _serviceEnabled;
    PermissionStatus _permissionGranted;
    LocationData _locationData;

    _serviceEnabled = await location.serviceEnabled();
    if (!_serviceEnabled) {
      _serviceEnabled = await location.requestService();
      if (!_serviceEnabled) return;
    }

    _permissionGranted = await location.hasPermission();
    if (_permissionGranted == PermissionStatus.denied) {
      _permissionGranted = await location.requestPermission();
      if (_permissionGranted != PermissionStatus.granted) return;
    }

    _locationData = await location.getLocation();
    location.onLocationChanged.listen((LocationData currentLocation) {
      widget.lastKnownLocation =
          LatLng(_locationData.latitude, _locationData.longitude);

      if (widget.onLocationUpdate != null)
        widget.onLocationUpdate(widget.lastKnownLocation);

      if (showMarker) {
        statefulMapController.addMarker(
            marker: Marker(
              point: widget.lastKnownLocation,
              builder: (ctx) => Container(
                child: Icon(
                  Icons.account_circle,
                  color: Colors.red.shade900,
                ),
              ),
            ),
            name: "currentLocation");

        statefulMapController.centerOnPoint(widget.lastKnownLocation);
      }
    });
  }
}
